### PR TITLE
[AutoDiff upstream] Store `@differentiable` original declaration.

### PR DESCRIPTION
**Pull request by dan-zheng**
Store in `DifferentiableAttr` the original declaration on which the attribute
is declared.

The original declaration is resolved during parsing and deserialization
(not yet upstreamed).

---

Progress towards TF-828: upstream `@differentiable` attribute type-checking.
`@differentiable` attribute type-checking mega-patch: https://github.com/apple/swift/pull/29091

---

Tests exist on `tensorflow` branch and will be upstreamed later.
Currently, not enough code has been upstreamed for meaningful testing.

### CODE REVIEW

**Review by rxwei — APPROVED** [review]

**Comment by dan-zheng** [comment]
@swift-ci Please test

File: lib/Parse/ParseDecl.cpp [hunk 1]
```diff
@@ -3473,6 +3473,18 @@ void Parser::setLocalDiscriminatorToParamList(ParameterList *PL) {
   }
 }
 
+/// Set the original declaration in `@differentiable` attributes.
+///
+/// Necessary beacuse `Parser::parseNewDeclAttribute` (which calls
```
Comment (dan-zheng) [inline lib/Parse/ParseDecl.cpp]:
```suggestion
/// Necessary because `Parser::parseNewDeclAttribute` (which calls
```

**Comment by dan-zheng** [comment]
@swift-ci Please test

**Comment by dan-zheng** [comment]
[Linux CI fails](https://ci.swift.org/job/swift-PR-Linux-smoke-test/19376/console) due to an unrelated error:
```
19:21:25 $ "/home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test/branch-master/buildbot_linux/none-swift_package_sandbox_linux-x86_64/usr/bin/swiftc" "-o" "/home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test/branch-master/buildbot_linux/none-swift_package_sandbox_linux-x86_64/tests/test-foundation-package/Output/test-foundation-xml.txt.tmp/test-foundation-xml" "/home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test/branch-master/swift-integration-tests/test-foundation-package/test-foundation-xml.swift"
19:21:25 # command stderr:
19:21:25 <unknown>:0: warning: missing submodule 'SwiftOverlayShims'
19:21:25 <unknown>:0: error: missing required module 'SwiftOverlayShims'
```
CI for other PRs seem to fail with the same error, so I imagine it's a recent breakage that should be fixed soon.

**Comment by dan-zheng** [comment]
@swift-ci Please smoke test Linux

**Comment by dan-zheng** [comment]
@swift-ci Please smoke test Linux platform

**Comment by dan-zheng** [comment]
@swift-ci Please smoke test

**Comment by dan-zheng** [comment]
Merging to unblock progress, happy to review additional feedback later!

**Pull request merged**